src/GafferImage/Catalogue.cpp

```diff
@@ -977,11 +977,29 @@ void Catalogue::imageRemoved( GraphComponent *graphComponent )
 		return;
 	}
 
-	Image *image = static_cast<Image *>( graphComponent );
-	// This causes the image to disconnect from
-	// the switch automatically.
-	removeChild( imageNode( image ) );
-	// So now we go through and shuffle everything down
+	// Remove the InternalImage corresponding to the external Image
+	// plug that was removed. We can't call `imageNode()` to find this
+	// for us, because the connections will have been removed along
+	// with the plug. So we search for an internal image which has
+	// no input connection.
+	bool removed = false;
+	for( auto &c : children() )
+	{
+		if( auto *internalImage = dynamic_cast<InternalImage *>( c.get() ) )
+		{
+			if( !internalImage->fileNamePlug()->getInput() )
+			{
+				assert( !removed );
+				// This causes the image to disconnect from
+				// the switch automatically.
+				removeChild( internalImage );
+				removed = true;
+			}
+		}
+	}
+	assert( removed ); (void)removed;
```

Comment:
I was puzzled by the void cast, though it wasn't too hard to look up online.  I guess the other option that might feel a little more self-documenting would be to put `bool removed = false;` and `removed = true;` in `#ifdef DEBUG`s, so that we don't declare the variable when we're not using it.

Reply:
I felt the `(void)` approach was preferable to breaking up the flow with three separate `#ifdef` sections. See what you think to https://github.com/GafferHQ/gaffer/pull/4620/commits/05f1242dfa4d6ef383f599bd04341fd021903237, which is hopefully a more self-explanatory approach using C++17's `[[maybe_unused]]` attribute.